pkg/controllers/routercerts/controller.go

```diff
@@ -194,26 +198,60 @@ func (c *routerCertsDomainValidationController) validateRouterCertificates() ope
 	// ignore the server cert from the default cert bundle
 	populateVerifyOptionsFromCertSlice(&verifyOptions, ingressCACerts)
 	if len(verifyOptions.Roots.Subjects()) == 0 { // the CA certs can also appear in the secret, but by default we should also trust the default ingress CA bundle for the default routes
-		return newRouterCertsDegradedf("NoRootCARouterCerts", "configMap/default-ingress-cert.data[ca-bundle.crt] -n openshift-config-managed: no root CA certificates found in the CM or system")
+		return []operatorv1.OperatorCondition{newRouterCertsDegradedf("NoRootCARouterCerts", "configMap/default-ingress-cert.data[ca-bundle.crt] -n openshift-config-managed: no root CA certificates found in the CM or system")}
 	}
 
 	serverCerts := populateVerifyOptionsFromCertSlice(&verifyOptions, certificates)
 	if len(serverCerts) == 0 {
-		return newRouterCertsDegradedf("NoServerCertRouterCerts", "secret/%v.spec.data[%v] -n %v: no server certificates found", c.defaultSecretName, ingressDomain, c.secretNamespace)
+		return []operatorv1.OperatorCondition{newRouterCertsDegradedf("NoServerCertRouterCerts", "secret/%v.spec.data[%v] -n %v: no server certificates found", c.defaultSecretName, ingressDomain, c.secretNamespace)}
 	}
 
 	// verify certificate chain
 	if err := verifyWithAnyCertificate(serverCerts, verifyOptions); err != nil {
-		return newRouterCertsDegradedf("InvalidServerCertRouterCerts", "secret/%v.spec.data[%v] -n %v: certificate could not validate route hostname %v: %v", c.defaultSecretName, ingressDomain, c.secretNamespace, verifyOptions.DNSName, err)
+		return []operatorv1.OperatorCondition{newRouterCertsDegradedf("InvalidServerCertRouterCerts", "secret/%v.spec.data[%v] -n %v: certificate could not validate route hostname %v: %v", c.defaultSecretName, ingressDomain, c.secretNamespace, verifyOptions.DNSName, err)}
 	}
 
 	// we made it this far without a problem
-	return operatorv1.OperatorCondition{
+	degradedCondition := operatorv1.OperatorCondition{
 		Type:   conditionRouterCertsDegradedType,
 		Status: operatorv1.ConditionFalse,
 		Reason: "AsExpected",
 	}
 
+	var upgradeCondition operatorv1.OperatorCondition
+	{
+		// check the server certificates for reliance on the deprecated CN field
+		var invalidCertificates []*x509.Certificate
+		for _, serverCert := range serverCerts {
+			if !crypto.CertHasSAN(serverCert) {
+				invalidCertificates = append(invalidCertificates, serverCert)
+			}
+		}
+		if len(invalidCertificates) > 0 {
+			message := fmt.Sprintf("the following certificates in 'secret/%v.spec.data[%v] -n %v' rely on legacy Common Name field, use SANs instead:", c.defaultSecretName, ingressDomain, c.secretNamespace)
```

Comment:
"the following are serial numbers of certificates in... that rely on..."

Reply:
ok

Reply:
ok; PTAL